### DISCUSSION TITLE
Floating point lists for NUOPC_CompAttributeGet

### DISCUSSION THREAD

**Discussion by NickSzapiro-NOAA** [esmf-org/esmf#291] [NUOPC] [answered]
### Requirements

- [X] Reviewed [NUOPC Reference Manual](https://earthsystemmodeling.org/doc/)
- [X] Searched [GitHub Discussions](https://github.com/orgs/esmf-org/discussions?discussions_q=)

### Affiliation(s)

Lynker in support of NCEP EMC

### ESMF Version

v8.6.1

### Issue

Hi. In short, I have floating point ALLCOMP attributes in a UFS configure file. Is it possible to add support for list of floating point to NUOPC_CompAttributeGet? 

I see that
- ESMF attributes support list of integer, real, double precision, logical, or character values
- NUOPC_CompAttributeGet is more limited to list of integer and string

 If not, I then need some intermediate conversion string-->double. Some current implementations of this are not robust across compiler options.
 
 Alternately, I don't know if there is some way to call something like ESMF_InfoGetAlloc but with a differently formatted UFS configure file... Any guidance/feedback is appreciated.

### Autotag

@danrosen25

**Comment by danrosen25** [comment, marked as answer]
Hi @NickSzapiro-NOAA 
We can add this to our backlog. Do you have a sense of the urgency?
I've created an issue [here](https://github.com/esmf-org/esmf/issues/292)
In the meantime this can be accomplished in user code using
`call ESMF_InfoGet(info, key="/NUOPC/Instance/"//name, values=valueList, rc=rc)`
but you'll want to add checks similar to how it's done in NUOPC for getting an integer list.
https://github.com/esmf-org/esmf/blob/edb848d90bf5d70a2db158771f4d07be605fdd32/src/addon/NUOPC/src/NUOPC_Comp.F90#L1324

**Comment by NickSzapiro-NOAA** [comment]
Thanks for the solution @danrosen25! I missed the "/NUOPC/Instance/" part for key and got stuck. 
I don't think this is urgent as we could edit implementation when available. It would be nice to have though, as there will be similar code in several components

**Comment by NickSzapiro-NOAA** [comment]
It looks like ESMF_InfoGetAlloc is misunderstanding the attribute type as string rather than number.
```
20240906 175245.904 ERROR            PET34 /scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.6.0/cache/build_stage/spack-stage-esmf-8.6.0-nnuwc5zlpvogeiuk3nec26eryjiwsopw/spack-src/src/Infrastructure/Base/src/ESMCI_Info.C:820 Info::get(): Failed type check (JSON trace will follow): /NUOPC/Instance/restart_fh
20240906 175245.904 ERROR            PET34 ESMCI_Info.C:822 Info::get() Attribute incorrect typekind  - [json.exception.type_error.302] type must be number, but is string
20240906 175245.904 ERROR            PET34 ESMCI_Info.C:836 Info::get() Attribute incorrect typekind  - Internal subroutine call returned Error
20240906 175245.904 ERROR            PET34 ESMC_InfoCDefGeneric.C:210 ESMC_InfoGetArrayR8() Attribute incorrect typekind  - Internal subroutine call returned Error
20240906 175245.904 ERROR            PET34 ESMF_Info.F90:1424 ESMF_InfoGetArrayR8Alloc() Attribute incorrect typekind  - Internal subroutine call returned Error
20240906 175245.904 ERROR            PET34 ice_comp_nuopc.F90:540 Attribute incorrect typekind  - Passing error in return code
20240906 175245.904 ERROR            PET34 UFS Driver Grid Comp:src/addon/NUOPC/src/NUOPC_Driver.F90:2901 Attribute incorrect typekind  - Phase 'IPDv01p1' Initialize for modelComp 4: ICE did not return ESMF_SUCCESS
20240906 175245.904 ERROR            PET34 UFS Driver Grid Comp:src/addon/NUOPC/src/NUOPC_Driver.F90:1324 Attribute incorrect typekind  - Passing error in return code
20240906 175245.904 ERROR            PET34 UFS Driver Grid Comp:src/addon/NUOPC/src/NUOPC_Driver.F90:486 Attribute incorrect typekind  - Passing error in return code
20240906 175245.904 ERROR            PET34 UFS.F90:394 Attribute incorrect typekind  - Aborting UFS
```
Can I set this somehow? I want array of type real(kind=ESMF_KIND_R8)

**Reply by danrosen25** [reply]
How are you setting the attribute / what does your configuration file look like for `restart_fh`? If you're using the slightly outdated ESMF_Config class then you should convert each item in a string list into real using function `num = ESMF_UtilString2Real(string, rc)`. I think the ESMF_Config class treats all input as a string. If you're using the ESMF_HConfig class (YAML) then remove any quotations from the `restart_fh` value.

**Reply by NickSzapiro-NOAA** [reply]
The idea is for user to be able to input something in UFS configure file like
https://github.com/NickSzapiro-NOAA/ufs-weather-model/blob/restart_fh_marine/tests/parm/ufs.configure.datm_cdeps.IN#L123
```
ALLCOMP_attributes::
...
      restart_n = 12
      restart_option = nhours
      restart_ymd = -999
      write_restart_at_endofrun = .true.
      restart_fh = 0.05,0.25,0.5,3,6
 ...
 ::
 ```
This keeps options together for components to write restarts. Can change delimiter/whitespace requirements, if it helps.

May I check that I understand? For a solution across ESMF versions and classes, input user's restart_fh as list of strings and call UtilString2Real on each?

In first testing, input shows up as one string with commas instead of array of strings. To split long string, subroutine decodeName looks great but not public

**Reply by danrosen25** [reply]
Hi @NickSzapiro-NOAA @theurich 
From what I can tell ESMF_Config -> ESMF_Info -> NUOPC Attributes isn't working correctly. There has been a focus on ESMF_HConfig -> ESMF_Info -> NUOPC Attributes recently, which might be why this has been missed.
**What's broken**
```
<COMP>_attributes::
      restart_fh = 0.05,0.25,0.5,3,6
 ::
```
Lists aren't working. This is seen as a single string.
```
<COMP>_attributes::
      restart_fh = 0.05, 0.25, 0.5, 3, 6
 ::
```
Spaces are breaking ESMF_Info. So NUOPC attributes doesn't store `restart_fh`at all. This is also true for `restart_fh = "0.05, 0.25, 0.5, 3, 6"`.

**Reply by danrosen25** [reply]
I've created a second issue for this bug [here](https://github.com/esmf-org/esmf/issues/293)
After some research I determined that NUOPC_CompAttributeIngest (called in the UFS driver to add attribute information to a component) fails silently if there are more than one item on the right hand side of the configuration line.

**Comment by NickSzapiro-NOAA** [comment]
@danrosen25 Maybe more fundamentally, may we ask if it would be possible for ESMF/NUOPC code to provide an 
`is_restart_time?-->true/false` function with model time input for all components to share?

We would want a slight tweak to restart_fh attribute above, where if second number is -1 then first number is interpreted as a frequency rather than specific forecast hour. This "encoding" is as in FV3 output_fh (https://github.com/NOAA-EMC/fv3atm/pull/345)

**Reply by danrosen25** [reply]
NUOPC doesn't have standard configuration options and each application customizes configuration. That is something address in UFS using [`ALLCOMP_attributes::`](https://github.com/ufs-community/ufs-weather-model/blob/2ccc549348da37aac51ab44482174dff2bb2912d/driver/UFSDriver.F90#L888). We're moving towards standard configuration options in [ESMX](https://github.com/esmf-org/esmf/tree/develop/src/addon/ESMX) and there are configuration options for the application but they're not passed to each component. It's yaml based and you can define variable and variable blocks.

**Reply by junwang-noaa** [reply]
@danrosen25 This is actually a more general question on ESMF clock. Right now ESMF clock provides info such as start time, current time, time step, etc. I am wondering if the clock can also provide restart time (which by nature applies to all the sub components) if a restart time info is provided in the top configuration file in ESMX. Once we set the restart time in the main driver, the restart time can be passed down to the subcomponents and an ESMF call (e.g. ESMF_isRestartTime) can be accessed from those sub components.  Currently we have to implement the restart info in every model components, which in my mind is not a clean solution.

**Reply by danrosen25** [reply]
@junwang-noaa What does it currently look like?
I think you're asking about attaching an [alarm](https://earthsystemmodeling.org/docs/release/latest/ESMF_refdoc/node6.html#SECTION06086400000000000000) to the driver clock. The driver clock is [accessible](https://earthsystemmodeling.org/docs/release/latest/NUOPC_refdoc/node4.html#SECTION00043100000000000000) from every NUOPC component. But I'm not sure if this is the solution you need. Are you asking to pass a time interval to every model and use that time interval to write restart files?

**Reply by junwang-noaa** [reply]
It is not actually alarm. Just like we can get currtime with starttime and time step/time interval, I'd like to see if ESMF can provide restarttime from ESMF clock. There is a limitation on using alarm as it has to be repeated in a fixed frequency. We actually would like to know the restart time and compare it directly with currtime without using alarm.

**Reply by danrosen25** [reply]
Hi @junwang-noaa,
I'll discuss this with the ESMF team on our core call. Thanks!

**Reply by danrosen25** [reply]
@junwang-noaa @NickSzapiro-NOAA @anntsay 
The ESMF Core team did some brainstorming and we have a resolution. Ann is going reach out to set up a planning discussion.

**Reply by anntsay** [reply]
Meeting note: 10/30:   

Jun:  trying to get restart time.. had to implement the configuration code in every component.. so hoping to have a flexible restart time =  a variable to integrate in the clock to help trigger restart/writing time.  current workaround, every component has a restart. -> 

Nick: current workaround implementation: [https://github.com/NickSzapiro-NOAA/ufs-weather-model/blob/restart_fh_marine/CDEPS-interface/ufs/cdeps_share/shr_is_restart_fh_mod.F90](https://github.com/NickSzapiro-NOAA/ufs-weather-model/blob/restart_fh_marine/CDEPS-interface/ufs/cdeps_share/shr_is_restart_fh_mod.F90)

Dan: clock cannot contain a clock and but can have an alarm. maybe have an alarm.

Jun: for example, restart every 12 hours.. so the component can check alarm and to trigger restart.  could have several alarms to mitigate different frequency of triggering.. but code is messy

Bill: possible to extend the current alarm capabilities.

Jun: question about alarm: subcomponent only inherit from parent component.. but somehow cannot see alarm passing down to subcomponent.    the original question is whether the restart time or alarm can be passed down to subcomponent
Dan-> can attach the alarm to the clock and has to manually done in the driver.
Gerhard -> fetchclock and set run clock -> set during run phase (in [NUOPC document](https://earthsystemmodeling.org/docs/nightly/develop/NUOPC_refdoc/node4.html#SECTION00043000000000000000))
Jun: the problem is that the alarm is a frequency 

Next Step:
1. UFS team is working Nick's current workaround 
2. another potential work around (Gerhard/Dan) - (1) Leverage reference time in the clock.  in Parent clock (driven by run sequence), could have the driver in charge, and has a list of time in a configuration file.   the driver register in the run sequence, and check current time and then look at the list of restart time and fetch that as a reference time.  can now the child to look up the reference time.  (2) use alarm 
3.  targeting 8.8  release -> trying to fix how the all_comp is working.
4. in future releases -> ESFM team will consider this feature request and work into clock/alarm class
